modules/dnn/src/layers/einsum_layer.cpp

```diff
@@ -0,0 +1,1176 @@
+    }
+    new_shape.assign(input_dims.begin(), input_dims.end());
+    for (size_t i = 0; i < perm.size(); ++i) {
+        new_shape[i] = input_dims[perm[i]];
+    }
+    return true;
+}
+
+Mat batchwiseMatMul(
+    const Mat& input1,
+    const MatShape& input1ShapeOverride,
+    const Mat& input2,
+    const MatShape& input2ShapeOverride)
+{
+
+    // Sanity checks before the actual MatMul
+    //input_1.DataType() == input_2.DataType(), "Data types of the inputs must match for MatMul");
+
+    CV_CheckEQ((size_t) input1ShapeOverride.size(), (size_t) 3, "Only 1 batch dimension is allowed for MatMul");
+    CV_CheckEQ((size_t) input2ShapeOverride.size(), (size_t) 3, "Only 1 batch dimension is allowed for MatMul");
+    CV_CheckEQ((size_t) input1ShapeOverride[0], (size_t) input2ShapeOverride[0], "Batch dimension should match for MatMul;");
+    CV_CheckEQ((size_t) input1ShapeOverride[2], (size_t) input2ShapeOverride[1], "Incompatible matrix dimensions for matMul");
+
+    size_t batches = input1ShapeOverride[0];
+    size_t M = input1ShapeOverride[1];
+    size_t K = input1ShapeOverride[2];
+    size_t N = input2ShapeOverride[2];
+
+    //TODO: deal with dynamic shapes
+    //TODO: deal with reshaping operation (it might not always be needed)
+    std::vector<Mat> output;
+    if (batches > 1)
+    {
+        Mat reshapedInput1 = input1;
+        Mat reshapedInput2 = input2;
+
+        // input1 should of size MxK
+        // check if input1 needs reshape, if need reshape
+        if (input1.size[0] != M || input1.size[1] != K)
+        {
+            int shape[] = {static_cast<int>(batches), static_cast<int>(M), static_cast<int>(K)};
+            reshapedInput1 = input1.reshape(1, 3, shape);
+        }
+
+        // input2 should be of size KxN
+        // check if input2 needs reshape, if needs reshape
+        if (input2.size[0] != K || input2.size[1] != N)
+        {
+            int shape[] = {static_cast<int>(batches), static_cast<int>(K), static_cast<int>(N)};
+            reshapedInput2 = input2.reshape(1, 3, shape);
+        }
+
+        for (size_t i=0; i < batches; i++)
+        {
+            std::vector<Range> ranges1 = {cv::Range(i, i+1)};
+            for (int j = 1; j < reshapedInput1.dims; j++)
+                ranges1.emplace_back(cv::Range::all());
+
+            Mat part1 = reshapedInput1(ranges1);
+            int shape[] = {static_cast<int>(M), static_cast<int>(K)};
+            part1 = part1.reshape(1, sizeof(shape)/sizeof(shape[0]), shape);
+
+            std::vector<Range> ranges2 = {cv::Range(i, i+1)};
+            for (int j = 1; j < reshapedInput2.dims; j++)
+                ranges2.emplace_back(cv::Range::all());
+
+            Mat part2 = reshapedInput2(ranges2);
+            int shape2[] = {static_cast<int>(K), static_cast<int>(N)};
+            part2 = part2.reshape(1, sizeof(shape2)/sizeof(shape2[0]), shape2);
+
+            Mat tmp_output;
+            cv::gemm(part1, part2, 1.0, cv::Mat(), 1.0, tmp_output);
```

Comment:
You can use the fast_gemm which is introduced in https://github.com/opencv/opencv/pull/23897 once it is merged.

Reply:
Or even entire Gemm layer, without using separate primitives.